python/ray/tests/test_concurrency_group.py

```diff
@@ -236,6 +240,39 @@ def test_tasks_on_different_executors(self, ray_start_regular_shared):
         assert value == "f2"
 
 
+def test_multiple_threads_in_same_group(ray_start_regular_shared):
+    """
+    This test verifies that all threads in the same concurrency group are still
+    alive from the Python interpreter's perspective even if Ray tasks have finished, so that
+    thread-local data will not be garbage collected.
+    """
+
+    @ray.remote
+    class Actor:
+        def __init__(self):
+            self.data = 0
+            self._thread_local_data = threading.local()
+
+        def set_thread_local(self, value: Any) -> int:
+            # If the thread-local data were garbage collected after the previous
+            # task on the same thread finished, `self.data` would be incremented
+            # more than once for the same thread.
+            if not hasattr(self._thread_local_data, "value"):
+                self._thread_local_data.value = self.data
+                self.data += 1
+            assert self._thread_local_data.value <= self.data
+
+        def get_data(self) -> int:
+            return self.data
+
+    max_concurrency = 5
+    a = Actor.options(max_concurrency=max_concurrency).remote()
+    for _ in range(200):
+        for i in range(max_concurrency):
+            ray.get(a.set_thread_local.remote(i))
+    assert ray.get(a.get_data.remote()) == max_concurrency
```

Comment:
the way the test is written relies on the assignment of tasks -> threads to be round robin. that's not a public/guaranteed behavior, so let's write the test in a way that's more defensive.

for example, you can use a counter or signal actor to block execution inside each method call and guarantee that in each batch of requests, 1 request is handled by each thread

the test also makes 1000 sequential actor calls currently, which is unnecessarily slow. i'd suggest just ending 2 batches of requests in a way similar to the above suggestion

Reply:
Updated test: https://github.com/ray-project/ray/pull/52575/commits/77879b71dff48d39eaabdb793b0de0a737a6b38e